jena-arq/src/main/java/org/apache/jena/riot/system/EltStreamRDFImpl.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.jena.riot.system;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+import org.apache.jena.graph.Triple;
+import org.apache.jena.riot.out.NodeFmtLib;
+import org.apache.jena.sparql.core.Quad;
+
+/** An item of a StreamRDF, including exceptions. Internal API. */
+public class EltStreamRDFImpl
+    implements EltStreamRDF, Serializable
+{
+    private static final long serialVersionUID = 1L;
+
+    Triple triple = null;
+    Quad quad = null;
+    String prefix = null; // Null implies "base".
+    String iri = null;
+    Throwable exception = null;
+
+    public EltStreamRDFImpl() {}
+
+    /* Prefer static constructors in EltStreamRDF */
+
+    public EltStreamRDFImpl(Triple triple) { this.triple = triple; }
+    public EltStreamRDFImpl(Quad quad) { this.quad = quad; }
+    public EltStreamRDFImpl(String prefix, String iri) { this.prefix = prefix; this.iri = iri; }
+    public EltStreamRDFImpl(String iri) { this.iri = iri; }
+    public EltStreamRDFImpl(Throwable exception) { this.exception = exception; }
+
+    @Override public boolean   isTriple()     { return triple != null; }
+    @Override public Triple    getTriple()    { return triple; }
+    @Override public boolean   isQuad()       { return quad != null; }
+    @Override public Quad      getQuad()      { return quad; }
+    @Override public boolean   isPrefix()     { return prefix != null; }
+    @Override public String    getPrefix()    { return prefix; }
+    @Override public boolean   isBase()       { return prefix == null && iri != null; }
+    @Override public String    getIri()       { return iri; }
+    @Override public boolean   isException()  { return exception != null; }
+    @Override public Throwable getException() { return exception; }
+
+    @Override
+    public EltStreamRDFType getType() {
```

Comment:
This class is really a tagged union value (immutable).

Suggestions
* have a slot for the type and set the type when constructed. 
* make all fields final.
* use `Objects.requireNonNull` in constructors to catch problems early. Then there is a guarantee one and only one slot is non-null.

Its tedious writing out the duplication but it is not visible.

And maybe it will be a Java `record` at Java17 (which also means it can't be an interface - see the implementation-of-one comment).

Reply:
> which also means it can't extends an interface

I guess I have to try that out myself - so far I haven't. From the snippets I find on the web I would have thought something like this might work:
```
interface Foo { String name(); }
record Bar (String name) implements Foo;
```
Its not clear to me whether bean properties (methods with set/get/is prefixes) would still auto-map to the record fields.

In any case, I can add your suggestions which would make EltStreamRDFImpl "clean" to use publicly.
Though there is then still the question whether the ctors should be public or whether construction should only be possible with the named static methods such as `EltStreamRDF.triple(t)` - rather than `new EltStreamRDF(t)`.

The intent of the interface was to decouple client code from these implementation considerations so that the implementation can evolve independently - but I am fine with removing it if it turns out that one implementation can be effectively authorative.

Reply:
The getters for records are `field()` style, no `get`. There are no setters because records are value-based/immutable.

I was wrong - records can implement interface- they can't extend anything.

Yes - this does seem to be authoritive as it is the output of the AsyncParser.

Reply:
> There are no setters because records are value-based/immutable.

Whoops, I mentioned setters just out of habit.

I replaced the EltStreamRDF interface with the implementation and attempted to make it 'record-like' for future-proofness, i.e. field() rather than getField(). I implemented all your suggestions.